motr/io_nw_xfer.c

```diff
@@ -1074,14 +1074,28 @@ static int target_ioreq_iofops_prepare(struct target_ioreq *ti,
 		rw_fop->crw_fid = ti->ti_fid;
 		rw_fop->crw_pver = ioo->ioo_pver;
 		rw_fop->crw_index = ti->ti_obj;
-		/* In case of partially spanned units in a parity group,
-		 * degraded read and read-verify mode expects zero-filled
-		 * units from server side.
+
+		/*
+		 * Use NOHOLE by default (i.e. return error for missing
+		 * units instead of zeros), unless we are in read-verify
+		 * mode or in the degraded-read mode. Otherwise, in case of
+		 * partially spanned parity groups (last groups, usually),
+		 * we will get a lot of bogus errors when all data units
+		 * of the group are read.
+		 *
+		 * Note: parity units are always present in the groups, even
+		 * in the partially spanned ones. So we always use NOHOLE
+		 * for them. Otherwise, the user may get corrupted data.
```

Comment:
I didn't remember how the following case was handled: partially spanned parity groups (last groups, usually) 

For a partially spanned group, the not-covered data units are not filled with zeroes?
Or those units are filled with zeroes only in client memory(in order to calculate parities) but not on ioservers?

Reply:
They are not filled with anything (to save the storage space). But in case of read-verify or degraded read - all units are read, so we must return zeros in this cases instead of errors (of non-existing units), from `stob_ad_read_prepare()`:

```
                if (seg->ee_val == AET_HOLE) {
                        if (io->si_flags & SIF_NOHOLE) {
                                rc = M0_ERR(-EIO);
                                break;
                        }
                        /*
                         * Read of a hole or unallocated space (beyond
                         * end of the file).
                         */
                        memset(stob_ad_addr_open(buf, bshift),
                               0, frag_size << bshift);
                        io->si_count += frag_size;
                }
```